File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
import sys


def solution():
    n, k = map(int, sys.stdin.readline().split())
    back_pack = [[]]
    dp = [[0]*(k+1) for _ in range(n+1)]

    for _ in range(n):
        back_pack.append(list(map(int, sys.stdin.readline().split())))

    for i in range(1, n+1):
        weight, value = back_pack[i]
        for j in range(1, k+1):
            if weight <= j:
                dp[i][j] = max(dp[i-1][j], dp[i-1][j-weight] + value)
            else:
                dp[i][j] = dp[i-1][j]

    print(dp[n][k])


solution()

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer tokenizer = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(tokenizer.nextToken());
        int m = Integer.parseInt(tokenizer.nextToken());

        int[][] backpack = new int[n + 1][2];
        for (int i = 1; i <= n; i++) {
            tokenizer = new StringTokenizer(br.readLine());

            int w = Integer.parseInt(tokenizer.nextToken());
            int v = Integer.parseInt(tokenizer.nextToken());

            backpack[i] = new int[] { w, v };
        }

        int[][] dp = new int[n + 1][m + 1];

        for(int i=1 ; i<=n ; i++){
            for(int weight=0 ; weight<=m ; weight++){
                if(backpack[i][0] > weight){
                    dp[i][weight] = dp[i - 1][weight];
                    continue;
                }

                dp[i][weight] = Math.max(dp[i - 1][weight], dp[i - 1][weight - backpack[i][0]] + backpack[i][1]);
            }
        }

        System.out.println(dp[n][m]);

        br.close();
    }
}